ruby/examples/lib/apply_order_discount.rb

```diff
@@ -0,0 +1,44 @@
+class ApplyOrderDiscount
+  Request = Struct.new(:id)
+  Response = Struct.new(:discount, :has_discount?)
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
+    @gateway = gateway
+  end
+
+  def apply(request)
+    order = gateway.find_order_by_id(request.id)
+
+    unless order.has_discount?
+      order.discount = discount_amount
+      gateway.save_order_discount(request.id, order)
+    end
+
+    Response.new(order.discount, order.has_discount?)
+  end
+
+  private
+
+  def discount_amount
+    10
+  end
+
+  class DiscountableOrder < Struct.new(:discount)
```

Comment:
No, I like the idea of keeping the domain objects extremely specific - this is the I in SOLID, until you know otherwise. The domain language should evolve as the requirements do.